server-controller/hooks_test.go

```diff
@@ -29,104 +29,444 @@ import (
+		newServer:    inferenceEndpoint,
+		expected:     "There is no such template",
+		clientErrors: nilClientError,
+	},
+	{
+		name:         "Success to update template",
+		oldServer:    inferenceEndpoint,
+		newServer:    inferenceEndpoint,
+		expected:     "updated successfully",
+		clientErrors: nilClientError,
+	},
+	{
+		name:      "Config map update error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during configMap update",
+		clientErrors: clientError{
+			configMapError:        mockClient{updateError: errorNew},
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Deployment update error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during deployment update",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClient{updateError: errorNew},
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Service deletion error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during service delete",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClient{deleteError: errorNew},
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Service creation error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during service create",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClient{createError: errorNew},
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Ingress patch error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during ingress update",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClient{updateError: errorNew},
+		},
+	},
+}
+
+func testUpdateTemplate(oldServer, newServer crv1.InferenceEndpoint, errs clientError) string {
+	var buf bytes.Buffer
+	log.SetOutput(&buf)
+	defer func() {
+		log.SetOutput(os.Stderr)
+	}()
+	updateMap := createTemplateClients(errs)
 	hooks := serverHooks{updateMap}
-	inferenceEndpoint := infer
-	hooks.Add(&inferenceEndpoint)
+	oldInferenceEndpoint := oldServer
+	newInferenceEndpoint := newServer
+	hooks.updateTemplate(&oldInferenceEndpoint, &newInferenceEndpoint)
 	return buf.String()
 }
 
+var patchDataTest = patchData{ModelName: "test", ModelVersionPolicy: "test", Namespace: "test", ResourcePath: "test"}
+
+var dataTestPrepareJSONPatchFromMap = []struct {
```

Comment:
These cases don't cover all paths in `prepareJSONPatchFromMap`, but in overall, these tests make 8% of overall `hooks` coverage, so they do test something in this function. I agree it could be done better with better mocking, but the goal of this PR was to achieve 50% coverage, and for now it is 60%. I think it could be done later.